python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll

```diff
@@ -0,0 +1,1540 @@
+   * Foo.method(foo, 42)
+   * ```
+   */
+  CallTypeMethodAsPlainFunction() or
+  /** A call to a class. */
+  CallTypeClass() or
+  /** A call on a class instance, that goes to the `__call__` method of the class */
+  CallTypeClassInstanceCall()
+
+/** A type of call. */
+class CallType extends TCallType {
+  string toString() {
+    this instanceof CallTypePlainFunction and
+    result = "CallTypePlainFunction"
+    or
+    this instanceof CallTypeNormalMethod and
+    result = "CallTypeNormalMethod"
+    or
+    this instanceof CallTypeStaticMethod and
+    result = "CallTypeStaticMethod"
+    or
+    this instanceof CallTypeClassMethod and
+    result = "CallTypeClassMethod"
+    or
+    this instanceof CallTypeMethodAsPlainFunction and
+    result = "CallTypeMethodAsPlainFunction"
+    or
+    this instanceof CallTypeClass and
+    result = "CallTypeClass"
+    or
+    this instanceof CallTypeClassInstanceCall and
+    result = "CallTypeClassInstanceCall"
+  }
+}
+
+// -------------------------------------
+// method call resolution
+// -------------------------------------
+private module MethodCalls {
+  /**
+   * Holds if `call` is a call to a method `target` on an instance or class, where the
+   * instance or class is not derived from an implicit `self`/`cls` argument to a method
+   * -- for that, see `callWithinMethodImplicitSelfOrCls`.
+   *
+   * It is found by making an attribute read `attr` with the name `functionName` on a
+   * reference to the class `cls`, or to an instance of the class `cls`. The reference the
+   * attribute-read is made on is `self`.
+   */
+  pragma[noinline]
+  private predicate directCall(
+    CallNode call, Function target, string functionName, Class cls, AttrRead attr, Node self
+  ) {
+    target = findFunctionAccordingToMroKnownStartingClass(cls, functionName) and
+    directCall_join(call, functionName, cls, attr, self)
+  }
+
+  /** Extracted to give good join order */
+  pragma[noinline]
+  private predicate directCall_join(
+    CallNode call, string functionName, Class cls, AttrRead attr, Node self
+  ) {
+    (
+      call.getFunction() = attrReadTracker(attr).asCfgNode() and
+      attr.accesses(classTracker(cls), functionName)
+      or
+      call.getFunction() = attrReadTracker(attr).asCfgNode() and
+      attr.accesses(classInstanceTracker(cls), functionName)
+    ) and
+    attr.accesses(self, functionName)
+  }
+
+  /**
+   * Holds if `call` is a call to a method `target` derived from an implicit `self`/`cls`
+   * argument to a method within the class `classWithMethod`.
+   *
+   * It is found by making an attribute read `attr` with the name `functionName` on a
+   * reference to an implicit `self`/`cls` argument. The reference the attribute-read is
+   * made on is `self`.
+   */
+  pragma[noinline]
+  private predicate callWithinMethodImplicitSelfOrCls(
+    CallNode call, Function target, string functionName, Class classWithMethod, AttrRead attr,
+    Node self
+  ) {
```

Comment:
Fair point. I think I've been too invested in the code to be able to see this any longer :sweat_smile: 

Looking over it now, I guess we _could_ turn `call` into a `DataFlow::CallCfgNode` (for all these functions). I just tried it out locally, and while it did work, I also didn't really find that it ended up giving any value from my point-of-view (it mostly removed a lot of the existing `.asCfgNode()` and introduced new ones in other places).

I think what happened is that I initially tried to keep everything on AST/controlf-flow layer, but moved things over to data-flow as needed for `resolveCall`/`getCallArg` (for example to be able to use synthetic data-flow nodes as arguments), and the rest just followed along for convenience.

I'm not really sure where to put a comment such as this, or whether it's any help, but happy to take suggestions :blush:

Reply:
Alright, this might be something we'd want to revisit when we've built out the data-flow layer a bit more (so all the coercion between layers hopefully isn't needed anymore).